numpy/core/src/multiarray/ctors.c

```diff
@@ -1183,28 +1183,43 @@ PyArray_NewFromDescrAndBase(
                                     flags, obj, base, 0, 0);
 }
 
-/*NUMPY_API
+/*
  * Creates a new array with the same shape as the provided one,
- * with possible memory layout order and data type changes.
+ * with possible memory layout order, data type and shape changes.
  *
  * prototype - The array the new one should be like.
  * order     - NPY_CORDER - C-contiguous result.
  *             NPY_FORTRANORDER - Fortran-contiguous result.
  *             NPY_ANYORDER - Fortran if prototype is Fortran, C otherwise.
  *             NPY_KEEPORDER - Keeps the axis ordering of prototype.
  * dtype     - If not NULL, overrides the data type of the result.
+ * ndim      - If not 0 and dims not NULL, overrides the shape of the result.
+ * dims      - If not NULL and ndim not 0, overrides the shape of the result.
  * subok     - If 1, use the prototype's array subtype, otherwise
  *             always create a base-class array.
  *
  * NOTE: If dtype is not NULL, steals the dtype reference.  On failure or when
  * dtype->subarray is true, dtype will be decrefed.
  */
 NPY_NO_EXPORT PyObject *
-PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
-                     PyArray_Descr *dtype, int subok)
+PyArray_NewLikeArrayWithShape(PyArrayObject *prototype, NPY_ORDER order,
+                              PyArray_Descr *dtype, int ndim, npy_intp *dims, int subok)
 {
     PyObject *ret = NULL;
-    int ndim = PyArray_NDIM(prototype);
+
+    int new_shape = 1;
+
+    if (dims == NULL) {
+        ndim = PyArray_NDIM(prototype);
+        dims = PyArray_DIMS(prototype);
+        new_shape = 0;
+    }
+    else if (order == NPY_KEEPORDER && (ndim != PyArray_NDIM(prototype) ||
+                !(PyArray_IS_C_CONTIGUOUS(prototype) || PyArray_IS_F_CONTIGUOUS(prototype)))) {
+        PyErr_SetString(PyExc_ValueError,
+                        "mismatching ndim or non-C/F-layout arrays can not keep order");
```

Comment:
Why are you checking `PyArray_IS_C_CONTIGUOUS` and `PyArray_IS_F_CONTIGUOUS` here?

Reply:
Because we can't keep the order when there's _no particular memory layout_, just like in the discussion here:
https://github.com/numpy/numpy/pull/13046#discussion_r272338815

Reply:
My current understanding of 'K' is that we try to keep the memory order _if possible_, not in all circumstances. Indeed, the former code fell back on recomputing the strides when the prototype was neither c/f continguous.

Reply:
First of all, we should make sure that this does not switch to an error suddenly. I agree that in most cases `K` probably means *if possible*.

I think the default may actually be `K` and not something like `None` (to mean `K if possible`), so that `K` is the most common option? If `K` was passed explicitly, there might be an argument.

There are some strides where it would be unambiguous, but I am not sure it is worth to bother about those.

Reply:
Unless I'm miserably misunderstanding some of the comments, I think that there are two clashing opinions:

1. Keep order _if_ possible (as my original implementation did), otherwise, fallback to something like C-contiguous (currently backed by @charris and @seberg);
1. Raise an exception if the order is `K` but not possible to maintain it (backed by @eric-wieser, previously discussed in https://github.com/numpy/numpy/pull/13046#discussion_r260673995).

Feel free to correct me if I'm really misunderstanding it, otherwise, it would be nice to have here @charris, @eric-wieser and @seberg commenting here to reach an agreement before I do any further changes regarding this matter.

Reply:
@pentschev yes, but sorry, I was a bit confused. I think the important thing is that:
```
np.ones_like(np.ones((10, 10, 10), order="F")[::2, :, ::2]).flags
```
does not raise an error, but since the shape is not changed it is OK in any case. So, I am actually OK with just giving an error, could go either way though.

Reply:
@seberg thanks for clarifying your concern. For my use case, both solutions are fine as well. At this point, I would just like to have a consensus to avoid modifying the PR back and forth.

As soon as there's a confirmation from both @charris and @eric-wieser we can move forward with the PR, either to merge it or to make changes as necessary.

Reply:
Just one more thing I thought about checking now, the [documentation on `numpy.array`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html) doesn't mention any errors/exceptions for order `K, saying only:

_F & C order preserved, otherwise most similar order_

So if there's an interest in maintaining the API closer to what happens in `numpy.array`, maybe raising an exception from these `*_like()` functions could be counterintuitive.

Reply:
> Because we can't keep the order when there's no particular memory layout,

My understanding is that we can and do - that's what the "sorted stride perm" function does

Reply:
My understanding from [this comment](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/shape.c#L814) is that it will do that provided that input arrays are C or F contiguous. For the condition we are talking about here, the array is neither. In other words, what does it mean to keep the order if it's not known?

Reply:
It means keep `argsort(strides[shape !=1])` unchanged (ignoring ties)

Reply:
I think I need to play around with that function, my interpretation might be wrong

Reply:
@eric-wieser could you let me know about your findings then? I'm sorry, but I don't understand what you mean with your previous statement, it may be getting a little too complex for my understand of NumPy handling of strides and orders.

Reply:
Sure, here's the case I'm thinking of:
```python
>>> a = np.zeros((20, 30, 40)).transpose((1, 0, 2))[::10,::10,::10]

>>> a.strides
(3200, 96000, 80)

>>> np.copy(a, order='K').strides
(32, 96, 8)
```

note how the order of the strides is kept.

This is the behavior I would expect for `empty_like` too

Reply:
@eric-wieser I believe the new commit I pushed covers now the case you're looking for. Extending the case from your previous comment, this is what we get with this new commit:

```python
>>> np.empty_like(a, shape=(3, 2, 4), order='K').strides
(32, 96, 8)
```

Can you confirm this is what you had in mind?

Reply:
That behavior looks correct to me now. Well take one more look at this tonight.

Reply:
@eric-wieser just checking, have you had the chance to have a look at this? Do you think we're good to merge this now?